compiler/modules/front/codeGen.bal

```diff
@@ -758,24 +804,97 @@ function codeGenIfElseStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
             }
             if elseContBlock != () {
                 elseContBlock.insns.push(branch);
-                assignments.push(...elseAssignments);
             }
-            return { block: contBlock, assignments };
+            assignments.push(...elseAssignments);
+            StmtNarrowing[] narrowings = combineIfElseNarrowings(ifNarrowings, ifContBlock != (), elseNarrowings, elseContBlock != ());
+            return { block: contBlock, assignments, narrowings };
+        }
+    }
+}
+
+function combineIfElseNarrowings(StmtNarrowing[] ifNarrowings, boolean ifCompletes, StmtNarrowing[] elseNarrowings, boolean elseCompletes) returns StmtNarrowing[] {
+    readonly & [boolean, boolean] pair = [ifCompletes, elseCompletes];
+    match pair {
+        [true, true] => {
+            return unionStmtNarrowings(ifNarrowings, elseNarrowings);
+        }
+        [true, false] => {
+            return ifNarrowings;
+        }
+        [false, true] => {
+            return elseNarrowings;
+        }
+        [false, false] => {
+            return [];
+        }
+    }
+    // JBUG
+    panic err:impossible();
```

Comment:
This will eventually cause our code not to compile so I would remove the `[false, false]` case above and return `[]` here.